storage/range_tree_test.go

```diff
@@ -167,343 +310,478 @@ func TestRotateRight(t *testing.T) {
-				Black:     false,
+				ParentKey: &keyNode,
+				LeftKey:   &keyLeftLeft,
 			},
 			expectedNode: &proto.RangeTreeNode{
-				Key:       keyLeft,
-				ParentKey: proto.KeyMin,
-				Black:     true,
-				RightKey:  &keyNode,
+				Key:       keyNode,
+				ParentKey: &keyLeft,
+				RightKey:  &keyRight,
 			},
 			expectedLeft: &proto.RangeTreeNode{
-				Key:       keyNode,
-				ParentKey: proto.KeyMin,
-				Black:     false,
+				Key:      keyLeft,
+				LeftKey:  &keyLeftLeft,
+				RightKey: &keyNode,
 			},
-			expectedErr: false,
 		},
 	}
 
 	for i, test := range testCases {
 		tc := createTreeContext(&keyNode, []*proto.RangeTreeNode{
+			test.parent,
 			test.node,
 			test.left,
+			test.leftRight,
 		})
 
 		// Perform the rotation.
-		actualNode, err := tc.rotateRight(test.node)
+		actualLeft, err := tc.rotateRight(test.node)
 		if err != nil {
-			// Did we expect the error?
-			if test.expectedErr {
-				continue
-			}
 			t.Fatal(err)
 		}
 
-		// Are both nodes dirty?
-		if !tc.nodes[string(keyNode)].dirty {
-			t.Errorf("%d: Expected node to be dirty", i)
-		}
-		if !tc.nodes[string(keyLeft)].dirty {
-			t.Errorf("%d: Expected left node to be dirty", i)
-		}
+		checkTreeNode(t, tc, i, "parent", keyParent, test.expectedParent, nil)
+		checkTreeNode(t, tc, i, "node", keyNode, test.expectedNode, nil)
+		checkTreeNode(t, tc, i, "left", keyLeft, test.expectedLeft, actualLeft)
+		checkTreeNode(t, tc, i, "leftRight", keyLeftRight, test.expectedLeftRight, nil)
 
-		// Is the Fetched Node the same as the retruend one?
-		actualNodeFetched, err := tc.getNode(&keyLeft)
+		// Perform the reverse rotation and expect to get the original values
+		// back.
+		actualNode, err := tc.rotateLeft(actualLeft)
 		if err != nil {
 			t.Fatal(err)
 		}
-		if !reflect.DeepEqual(actualNodeFetched, actualNode) {
-			t.Errorf("%d: Fetched Node is not the same as the returned one. Fetched: %+v Returned: %+v", i, actualNodeFetched, actualNode)
-		}
-
-		// Are the resulting nodes correctly rotated?
-		if !reflect.DeepEqual(actualNode, test.expectedNode) {
-			t.Errorf("%d: Expected Node is not the same as the actual.\nExpected: %+v\nActual: %+v\n", i, test.expectedNode, actualNode)
-		}
 
-		actualLeft, err := tc.getNode(&keyNode)
-		if err != nil {
-			t.Fatal(err)
-		}
-		if !reflect.DeepEqual(actualLeft, test.expectedLeft) {
-			t.Errorf("%d: Expected Left Node is not the same as the actual.\nExpected: %+v\nActual: %+v\n", i, test.expectedLeft, actualLeft)
-		}
+		checkTreeNode(t, tc, i, "parent-reversed", keyParent, test.parent, nil)
+		checkTreeNode(t, tc, i, "node-reversed", keyNode, test.node, actualNode)
+		checkTreeNode(t, tc, i, "left-reversed", keyLeft, test.left, nil)
+		checkTreeNode(t, tc, i, "leftRight-reversed", keyLeftRight, test.leftRight, nil)
 	}
 }
 
-// TestRotateLeft ensures that left rotations occur correctly.
+// TestRotateLeft ensures that right rotations occur correctly. This is a
+// mirror of TestRotateRight.
```

Comment:
These rotation tests were extremely difficult to read for the amount that they test; the tree structure being tested is *not* clear.